### PR TITLE
Add grid navigation

### PR DESCRIPTION
**Pull request by goremikins**

### CODE REVIEW

File: packages/utils/src/grid/navigation.ts [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+ * 1 1 0 1
+ * 1 0 1 1
+ * 1 1 0 1
+ */
+type NavigationGrid = NavigationRow[];
+
+/**
+ * Cell, where first number is row index and second number is column index
+ */
+type CellIndex = [number, number];
```
[the remark this answers is not in the document]
Comment (goremikins) [reply packages/utils/src/grid/navigation.ts:20]: Done

File: packages/utils/src/grid/navigation.ts [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+ * @param grid Navigation grid
+ * @returns cell The last active cell. If none return null
+ */
+const lastCell = (grid: NavigationGrid): CellIndex | null => {
+  const [rowCount, columnCount] = counts(grid);
+  return previousCell(grid, [rowCount - 1, columnCount]);
+};
+
+export {
+  NavigationRow,
```
[the remark this answers is not in the document]
Comment (goremikins) [reply packages/utils/src/grid/navigation.ts:181]: No need to export separately as TypeScript does not care.

File: packages/utils/src/grid/navigation.ts [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+  NavigationGrid,
+  CellIndex,
+  nextCell,
```
[the remark this answers is not in the document]
Comment (goremikins) [reply packages/utils/src/grid/navigation.ts]: Methods have been renamed to more user friendly.

File: packages/utils/src/grid/navigation.ts [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+ * find the closest cell on the preceding rows
+ * @param grid Navigation grid
+ * @param cell Start cell
+ * @returns cell Closest cell on the previous row. If none return null
+ */
+const up = (grid: NavigationGrid, cell: CellIndex): CellIndex | null => {
+  const columnIndex = cell[0];
+  let rowIndex = cell[1];
+
+  while ((rowIndex -= 1) >= 0) {
```
Comment (TremayneChrist) [inline packages/utils/src/grid/navigation.ts:131]:
```suggestion
  for (let rowIndex = cell[1]; rowIndex >= 0; rowIndex -= 1) {
```

File: packages/utils/src/grid/navigation.ts [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+ * @param grid Navigation grid
+ * @param cell Start cell
+ * @returns cell Closest cell on the next row. If none return null
+ */
+const down = (grid: NavigationGrid, cell: CellIndex): CellIndex | null => {
+  const columnIndex = cell[0];
+  let rowIndex = cell[1];
+  const [rowCount] = counts(grid);
+
+  while ((rowIndex += 1) < rowCount) {
```
Comment (TremayneChrist) [inline packages/utils/src/grid/navigation.ts:153]:
```suggestion
  for (let rowIndex = cell[1]; rowIndex < rowCount; rowIndex += 1) {
```

File: packages/utils/src/grid/navigation.ts (hunk 4 above)
Comment (goremikins) [reply packages/utils/src/grid/navigation.ts:131]: It is not the same, the same is `for (let rowIndex = cell[1] - 1; rowIndex >= 0; rowIndex -= 1) {`, and this is more difficult to read.

File: packages/utils/src/grid/navigation.ts [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+ * Indexes are zero-based
+ */
+type CellIndex = [number, number];
+
+/**
+ * Get number of rows and columns in the grid
+ * @param grid Navigation grid
+ * @returns counts RowCount and ColumnCount
+ */
+const counts = (grid: NavigationGrid): [number, number] => [grid.length, grid.length ? grid[0].length : 0];
```
[the remark this answers is not in the document]
Comment (goremikins) [reply packages/utils/src/grid/navigation.ts]: OK, done

File: packages/utils/src/grid/navigation.ts (hunk 4 above)
Comment (TremayneChrist) [reply packages/utils/src/grid/navigation.ts:131]: Yeah I missed the `-1`. Anyway it's fine, it's just you're wrapping the assign to meet the linting rules.

File: packages/utils/src/grid/navigation.ts [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+ * Navigate down from the start cell trying to
+ * find the closest cell on the following rows
+ * @param grid Navigation grid
+ * @param cell Start cell
+ * @returns cell Closest cell on the next row. If none return null
+ */
+const down = (grid: NavigationGrid, cell: CellIndex): CellIndex | null => {
+  const columnIndex = cell[0];
+  let rowIndex = cell[1];
+  const rowCount = counts(grid)[1];
```
Comment (TremayneChrist) [inline packages/utils/src/grid/navigation.ts:151]: If you want to keep it the same as the others, I think you can do
```suggestion
  const [,rowCount] = counts(grid);
```

Comment (goremikins) [reply packages/utils/src/grid/navigation.ts]: Done

**Review by TremayneChrist — APPROVED** [review]

**Pull request merged**